tests/app/oidc/test_gcp_oidc.py

```diff
@@ -0,0 +1,29 @@
+from mock import patch
+
+from app.oidc.gcp_oidc import OIDCCredentialsServiceGCP
+
+TEST_SDS_OAUTH2_CLIENT_ID = "TEST_SDS_OAUTH2_CLIENT_ID"
+
+
+@patch("app.oidc.gcp_oidc.Request")
+@patch("app.oidc.gcp_oidc.fetch_id_token_credentials")
+def test_oidc_credentials_service_gcp_ttl(mock_token_fetch, mock_request):
+    oidc_credentials_service = OIDCCredentialsServiceGCP()
+
+    # fetch credentials and check the gcp call is made
+    oidc_credentials_service.get_credentials(iap_client_id=TEST_SDS_OAUTH2_CLIENT_ID)
+    mock_token_fetch.assert_called_once_with(
+        audience=TEST_SDS_OAUTH2_CLIENT_ID, request=mock_request()
+    )
+    mock_token_fetch.return_value.refresh.assert_called_once_with(mock_request())
+
+    # credentials are cached
+    oidc_credentials_service.get_credentials(iap_client_id=TEST_SDS_OAUTH2_CLIENT_ID)
+    assert mock_token_fetch.call_count == 1
+    assert mock_token_fetch.return_value.refresh.call_count == 1
+
+    # re-fetched when cache expires
+    oidc_credentials_service.get_credentials.cache_clear()
```

Comment:
We should have a test with mocked TTLs, this current one asserts that the cache works as expected if the cache is cleared, but this line is not something we execute. So if the lib had a bug and TTLs weren't expiring this test would still pass, but credentials would end up as expired and not refresh.

Reply:
I ran into some issues with this one, we can't mock time with freezegun because the TTLCache has a cached version of the time function, I've replaced the timer with something that is updated by freezegun and checks that the cache behaves as we want it to at the 30, 55 min, 109 min, 110 min markers, let me know if thats the sort of thing you were thinking/thoughts

Reply:
Exactly what I was looking for, nice one. The custom ttl wrapper isn't great, but that is fine given it is an issue with the lib and not in our control.